### PR TITLE
Better serialization for ranges, domains, and iterators in non-default formats

### PR DESCRIPTION
**Pull request by benharsh**
This PR improves the serialization of ranges, domains, and iterators in non-default formats. The formatting choices here will be considered "unstable" for the upcoming release. That said, users can at least read and write these types in a way that is compatible with a given serialization format, rather than having the default format appear unexpectedly in the middle of a serialized object.

Post-2.0: determine the appropriate representation of these types through the serialization API.

Testing:
- [x] local
- [x] gasnet

### CODE REVIEW

**Comment by benharsh** [comment]
Tracking issue created here: https://github.com/chapel-lang/chapel/issues/23394

File: modules/internal/DefaultRectangular.chpl [hunk 1]
```diff
@@ -1717,6 +1717,26 @@ module DefaultRectangular {
     rwLiteral("}");
   }
 
+  proc DefaultRectangularDom.dsiSerialWrite(f) throws
+  where _supportsSerializers(f) && f.serializerType != IO.defaultSerializer {
+    if chpl_warnUnstable then
+      compilerWarning("Serialization of rectangular domains with non-default Serializer is unstable, and may change in the future");
```
Comment (jeremiah-corrado) [inline modules/internal/DefaultRectangular.chpl:1723]: I think moving this to an `@unstable` annotation, might remove some of the stack trace through the IO internals from the warning output. (I believe there is some special logic for `@unstable` to do that, but I could be wrong).

File: modules/standard/ChapelIO.chpl [hunk 2]
```diff
@@ -798,6 +798,30 @@ module ChapelIO {
...
+    }
+  }
+
+  @chpldoc.nodoc
+  proc _iteratorRecord.serialize(writer, ref serializer) throws {
+    if serializer.type == IO.defaultSerializer {
+      writeThis(writer);
+    } else {
+      if chpl_warnUnstable then
+        compilerWarning("Serialization of iterators with non-default Serializer is unstable, and may change in the future");
```
Comment (jeremiah-corrado) [inline modules/standard/ChapelIO.chpl:818]: For this one and for ranges, I guess you'd need to annotate a helper method.

**Review by jeremiah-corrado — APPROVED** [review]
Looks good!

One thought about reducing the noise of the unstable warnings:

File: modules/internal/DefaultRectangular.chpl (hunk 1 above)
Comment (benharsh) [reply modules/internal/DefaultRectangular.chpl:1723]: I tried doing this initially here and in the other places, but the attribute wasn't causing any warnings with ``--warn-unstable``. Maybe something to do with being called from the standard library?

Comment (jeremiah-corrado) [reply modules/internal/DefaultRectangular.chpl:1723]: Oh true, I guess they are suppressed in standard modules.

**Pull request merged**